request-manager/graph/grapher.go

```diff
@@ -207,78 +238,93 @@ func buildSeqGraph(seqSpec *spec.Sequence, idgen id.Generator) (seqGraph *Graph,
 			Name: nodeSpec.Name,
 			Spec: nodeSpec,
 		}
-		nodes[nodeSpec.Name] = &Graph{
+		g := &Graph{
 			Name:     n.Name,
 			Source:   n,
 			Sink:     n,
 			Nodes:    map[string]*Node{n.Id: n},
 			Edges:    map[string][]string{},
 			RevEdges: map[string][]string{},
 		}
+		nodes[nodeSpec.Name] = g
+		nodesToAdd[nodeSpec.Name] = g
 	}
 
-	// Nodes we've yet to add
-	nodesToAdd := map[string]*Graph{}
-	for k, v := range nodes {
-		nodesToAdd[k] = v
-	}
 	nodesAdded := map[string]bool{}
 
 	// Build graph by adding nodes, starting from the source node, and then
 	// adding all adjacent nodes to the source node, and so on.
-	// We cannot add nodes in any order because we do not know the reverse dependencies
+	// Nodes are not specified in any particular order in the specs; the
+	// only ordering information we have is from the `deps` field for each
+	// node. So, we continuously loop through the spec nodes, and add it to
+	// the graph only once all its dependencies have been added.
 	for len(nodesToAdd) > 0 {
 
-		componentAdded := false
-
-		for nodeName, component := range nodesToAdd {
+		// True iff we find a new node to add in this iteration. Assume
+		// false until we find such a node.
+		nodeAdded := false
+
+		// Each loop through the nodes, we should be able to add at least
+		// one node. If not, the there's a cycle (or perhaps a typo in
+		// deps: that wasn't caught earlier). For example, with A->B->C,
+		// the loop will add A since it has no deps, then it'll add B since
+		// it depends on A which has been built, then C. But if a cycle
+		// existed between B and C, the second loop wouldn't be able to
+		// build B and nodeAdded would be false and trigger the error
+		// after this loop.
+		for nodeName, node := range nodesToAdd {
 			nodeSpec := seqSpec.Nodes[nodeName]
-			if dependenciesSatisfied(nodesAdded, nodeSpec.Dependencies) {
-				// Dependencies for node have been satisfied; presumably, all input job args are
-				// present in job args map. If not, it's an error.
-				missingArgs, err := getMissingArgs(nodeSpec, jobArgs)
+			if !dependenciesSatisfied(nodesAdded, nodeSpec.Dependencies) {
```

Comment:
I think s/dependenciesSatisfied/haveAllDeps/ makes the func more intention revealing. "Satisfied" is also anthropomorphic. Software has no emotions. :-) What we really mean here is: we can't build this node until we have (built) all its dependencies.